torch/_inductor/memory.py

```diff
@@ -0,0 +1,533 @@
+                nodes_to_schedule.add(succ_node)
+
+        # update predecessor nodes
+        for buf in selected_node.mpi.pred_buffers:
+            assert buf.mpi.outdegree > 0
+            buf.mpi.outdegree -= 1
+            if buf.mpi.outdegree == 1:
+                for succ_node in buf.mpi.succ_nodes:
+                    succ_node.mpi.memory_to_free += buf.mpi.size_free
+
+    if num_iters > len(nodes):
+        raise RuntimeError("Failed to schedule, while loop ran too long for lpmf")
+
+    return schedule
+
+
+def topological_sort_bfs(nodes: List[BaseSchedulerNode]) -> List[BaseSchedulerNode]:
+    """
+    A BFS topological sort that selects nodes whose dependencies are executed
+    the earliest. This follows a FIFO idea.
+    """
+
+    @dataclasses.dataclass
+    class HeapElement:
+        priority: List[int]
+        node: BaseSchedulerNode
+
+        def __lt__(self, other: HeapElement) -> bool:
+            if self.priority == other.priority:
+                return self.node.mpi.index < other.node.mpi.index
+            return self.priority < other.priority
+
+    def _node_priority(node: BaseSchedulerNode) -> List[int]:
+        assert node.mpi.indegree == 0
+        ids = sorted({pred_node.mpi.index for pred_node in node.mpi.pred_nodes})
+        ids.append(node.mpi.index)
+        return ids
+
+    # compute nodes' number of unmet dependencies (for schedulability)
+    # initialize the list of nodes ready to be scheduled
+    nodes_to_schedule: List[HeapElement] = []
+    for t, node in enumerate(nodes):
+        node.mpi.index = t
+        # note that .unmet_dependencies could have deps with the same name
+        # and in that case, it should only be counted once
+        node.mpi.indegree = len(node.mpi.pred_nodes)
+        if node.mpi.indegree == 0:
+            heapq.heappush(nodes_to_schedule, HeapElement(_node_priority(node), node))
+
+    # schedule nodes one at a time
+    schedule: List[BaseSchedulerNode] = []
+    num_iters: int = 0
+    while num_iters < len(nodes) and nodes_to_schedule:
+        # select a node to schedule
+        selected_node = heapq.heappop(nodes_to_schedule).node
+        selected_node.mpi.index = len(schedule)
+        schedule.append(selected_node)
+        num_iters += 1
+
+        # update successor nodes and nodes_to_schedule
+        for succ_node in selected_node.mpi.succ_nodes:
+            assert succ_node.mpi.indegree > 0
+            succ_node.mpi.indegree -= 1
+            if succ_node.mpi.indegree == 0:
+                heapq.heappush(
+                    nodes_to_schedule,
+                    HeapElement(_node_priority(succ_node), succ_node),
+                )
+    if num_iters > len(nodes):
+        raise RuntimeError("Failed to schedule, while loop ran too long for bfs")
+    return schedule
+
+
+def topological_sort_dfs(nodes: List[BaseSchedulerNode]) -> List[BaseSchedulerNode]:
+    """
+    Ensure nodes is in topologically sorted order
```

Comment:
would you like to add more details on how this algorithm works, and some intuitive understanding on why it can lower peak memory? (would be great if we can have similar level of detail as in `topological_sort_lpmf`)